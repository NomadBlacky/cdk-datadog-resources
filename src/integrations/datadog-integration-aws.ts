import { CfnResource } from 'aws-cdk-lib';
import { Construct } from 'constructs';
// eslint-disable-next-line @typescript-eslint/no-require-imports
import camelcaseKeys = require('camelcase-keys');
import { DatadogCredentials } from '../common/properties';

export interface DatadogIntegrationAWSProps {
  /** Credentials for the Datadog API */
  readonly datadogCredentials: DatadogCredentials;

  /** The id of the account with which to integrate. */
  readonly accountId: string;

  /** The name of the created role. */
  readonly roleName: string;
}

/**
 * Datadog Integration 1.1.0
 */
export class DatadogIntegrationAWS {
  constructor(scope: Construct, id: string, props: DatadogIntegrationAWSProps) {
    const cfnProperties = camelcaseKeys(props, {
      deep: true,
      pascalCase: true,
    });
    // @ts-ignore
    delete Object.assign(cfnProperties, { AccountID: cfnProperties.AccountId }).AccountId;
    new CfnResource(scope, id, {
      type: 'Datadog::Integrations::AWS',
      properties: { ...cfnProperties },
    });
  }
}
