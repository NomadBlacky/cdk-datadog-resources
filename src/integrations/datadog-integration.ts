import { CfnResource, Construct } from '@aws-cdk/core';
import { DatadogCredentials } from '../common/properties';

export interface DatadogIntegrationProps {
  /** Credentials for the Datadog API */
  readonly datadogCredentials: DatadogCredentials;

  /** The id of the account with which to integrate. */
  readonly accountId: string;

  /**
   * The name of the created role.
   *
   * @default DatadogIntegrationRole
   * */
  roleName?: string;

}

/**
 * Datadog Integration 1.0.0
 */
export class DatadogIntegration {
  constructor(scope: Construct, id: string, props: DatadogIntegrationProps) {
    new CfnResource(scope, id, {
      type: 'Datadog::Integrations::AWS',
      properties: {
        DatadogCredentials: {
          ApiKey: props.datadogCredentials.apiKey,
          ApplicationKey: props.datadogCredentials.applicationKey,
        },
        AccountID: props.accountId,
        RoleName: props.roleName || 'DatadogIntegrationRole',
      },
    });
  }
}