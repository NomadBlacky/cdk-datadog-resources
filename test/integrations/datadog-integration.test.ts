import { SynthUtils } from '@aws-cdk/assert';
import '@aws-cdk/assert/jest';
import { Stack } from '@aws-cdk/core';
import { DatadogIntegration } from '../../src/integrations/datadog-integration';

test('Snapshot test', () => {
  const stack = new Stack();

  new DatadogIntegration(stack, 'TestIntegration', {
    datadogCredentials: {
      apiKey: 'DATADOG_API_KEY',
      applicationKey: 'DATADOG_APP_KEY',
    },
    accountId: '123456',
    roleName: 'DatadogAWSAcctRoleName',
  });

  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});

test('Default role name is assigned', () => {
  const stack = new Stack();

  new DatadogIntegration(stack, 'TestIntegration', {
    datadogCredentials: {
      apiKey: 'DATADOG_API_KEY',
      applicationKey: 'DATADOG_APP_KEY',
    },
    accountId: '123456',
  });

  expect(stack).toHaveResource('Datadog::Integrations::AWS', {
    RoleName: 'DatadogIntegrationRole',
  });
});
