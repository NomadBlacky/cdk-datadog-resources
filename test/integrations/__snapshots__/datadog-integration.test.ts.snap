// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test 1`] = `
Object {
  "Resources": Object {
    "TestIntegration": Object {
      "Properties": Object {
        "AccountID": "123456",
        "DatadogCredentials": Object {
          "ApiKey": "DATADOG_API_KEY",
          "ApplicationKey": "DATADOG_APP_KEY",
        },
        "RoleName": "DatadogAWSAcctRoleName",
      },
      "Type": "Datadog::Integrations::AWS",
    },
  },
}
`;
